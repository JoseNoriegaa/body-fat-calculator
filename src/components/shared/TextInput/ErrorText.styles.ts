// External dependencies
import { createUseStyles } from 'react-jss';

const useStyles = createUseStyles({
  error: {
    fontSize: '12px',
    fontWeight: 200,
    margin: 0,
    padding: 0,
    color: 'red',
  },
});

export default useStyles;
