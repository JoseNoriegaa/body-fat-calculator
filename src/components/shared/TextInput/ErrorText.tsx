// External dependencies
import React from 'react';

// Internal dependencies
import useStyles from './ErrorText.styles';

function ErrorText({ children }: React.PropsWithChildren<unknown>) {
  const classes = useStyles();
  return <p className={classes.error}>{children}</p>;
}

export default ErrorText;
