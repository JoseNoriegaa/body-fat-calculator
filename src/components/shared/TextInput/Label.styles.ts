// External dependencies
import { createUseStyles } from 'react-jss';

const useStyles = createUseStyles({
  label: {
    fontWeight: 200,
    fontSize: '14px',
    margin: '8px 0',
  },
});

export default useStyles;
