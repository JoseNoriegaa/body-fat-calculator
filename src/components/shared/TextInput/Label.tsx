/* eslint-disable jsx-a11y/label-has-associated-control */
// External dependencies
import React from 'react';

// Internal dependencies
import useStyles from './Label.styles';

// Types & Interfaces
interface ILabelProps {
  htmlFor?: string;
}

function Label({ children, htmlFor }: React.PropsWithChildren<ILabelProps>) {
  const classes = useStyles();

  return (
    <label htmlFor={htmlFor} className={classes.label}>
      {children}
    </label>
  );
}

export default Label;
