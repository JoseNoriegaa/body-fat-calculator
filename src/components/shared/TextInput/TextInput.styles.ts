// External dependencies
import { createUseStyles } from 'react-jss';

// Internal dependencies

const useStyles = createUseStyles({
  container: {
    display: 'flex',
    flexDirection: 'column',
    position: 'relative',
  },
  input: {
    borderWidth: '1px',
    borderStyle: 'solid',
    borderColor: 'rgba(255,255,255, 0.3)',
    height: '30px',
    borderRadius: '30px',
    backgroundColor: 'unset',
    padding: '0 15px',
    verticalAlign: 'middle',
    color: 'unset',
    transition: 'border-color .2s ease-out',
    '&:focus-visible': {
      outline: 'none',
      borderColor: 'rgba(255,255,255, .5)',
      backgroundColor: 'rgba(255,255,255,0.05)',
    },
  },
  inputError: {
    borderColor: 'red',
    '&:focus-visible': {
      borderColor: 'red',
    },
  },
  errorContainer: {
    position: 'absolute',

    bottom: '-4px',
    transform: 'translateY(100%)',
  },
});

export default useStyles;
