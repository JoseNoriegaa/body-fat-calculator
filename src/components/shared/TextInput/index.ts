export { default } from './TextInput';
export { default as Label } from './Label';
export { default as ErrorText } from './ErrorText';
