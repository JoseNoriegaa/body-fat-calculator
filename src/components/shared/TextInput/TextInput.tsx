// External dependencies
import React from 'react';
import clsx from 'clsx';

// Internal dependencies
import Label from './Label';
import useStyles from './TextInput.styles';
import ErrorText from './ErrorText';

// Types & Interfaces
interface ITextInputProps {
  name: string;
  placeholder: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  value?: string | number;
  label?: string;
  error?: string;
}

function TextInput({ onChange, name, value, placeholder, label, error }: ITextInputProps) {
  const classes = useStyles();

  const inputId = `input_${name}`;

  return (
    <div className={classes.container}>
      {label && <Label htmlFor={inputId}>{label}</Label>}

      <input
        id={inputId}
        type="text"
        name={name}
        className={clsx(classes.input, error && classes.inputError)}
        placeholder={placeholder}
        value={value}
        onChange={onChange}
      />

      {error && (
        <div className={classes.errorContainer}>
          <ErrorText>{error}</ErrorText>
        </div>
      )}
    </div>
  );
}

export default TextInput;
