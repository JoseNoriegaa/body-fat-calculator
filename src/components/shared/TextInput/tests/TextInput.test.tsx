// External dependencies
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';

// Internal dependencies
import TextInput from '../TextInput';

// Types & Interfaces
type TTextInputProps = React.ComponentProps<typeof TextInput>;

// Mocks
jest.mock('../ErrorText.tsx', () => ({
  __esModule: true,
  default: () => <span>error</span>,
}));

jest.mock('../Label.tsx', () => ({
  __esModule: true,
  default: () => <span>label</span>,
}));

const onChangeMock = jest.fn();

// Constants
const makeProps = (props?: Partial<TTextInputProps>) => {
  const output = {
    name: 'form_field',
    onChange: onChangeMock,
    placeholder: 'placeholder',
    error: undefined,
    label: 'Form Field',
    value: '',
    ...props,
  } as TTextInputProps;

  return output;
};

describe('<TextInput />', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<TextInput {...makeProps()} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the Label component if the label property was provided', () => {
    render(<TextInput {...makeProps()} />);

    expect(screen.getByText('label')).toBeInTheDocument();
  });

  it('should render the ErrorText component if the error property was provided', () => {
    render(<TextInput {...makeProps({ error: 'something went wrong' })} />);

    expect(screen.getByText('error')).toBeInTheDocument();
  });

  it('should handle onChange', () => {
    render(<TextInput {...makeProps({ value: 'input value' })} />);

    const input = screen.getByDisplayValue('input value');
    expect(input).toHaveProperty('id', 'input_form_field');

    fireEvent.change(input, { target: { value: 'new text' } });

    expect(onChangeMock).toHaveBeenCalled();
  });
});
