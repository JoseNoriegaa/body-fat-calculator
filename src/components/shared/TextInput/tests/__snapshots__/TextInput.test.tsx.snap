// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container-0-2-1"
  >
    <span>
      label
    </span>
    <input
      class="input-0-2-2"
      id="input_form_field"
      name="form_field"
      placeholder="placeholder"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
